### PR TITLE
fix(core): fix ldml transforms 🙀

### PR DESCRIPTION
**Pull request by srl295**
- document transform calling code in ldml_transform
- remove obsolete portions
- optimize keystroke code when no transforms are present
- also, fix 'enter' and 'delete' breakage, and other keys passed to the OS like 'tab'

@keymanapp-test-bot skip

#7377

### CODE REVIEW

**Comment by srl295** [comment]
https://github.com/keymanapp/keyman/assets/855219/c2c335c8-c78b-4d7a-aab2-ca735c31b7ee

**Comment by srl295** [comment]
Also fix backspace.

File: core/src/ldml/ldml_processor.cpp [hunk 1]
```diff
@@ -225,77 +225,84 @@ ldml_processor::process_event(
           state->actions().push_backspace(KM_KBP_BT_UNKNOWN);
         } else {
           state->actions().push_backspace(KM_KBP_BT_CHAR, last_char);
+          state->context().pop_back();
         }
-        state->context().pop_back();
```
Comment (srl295) [inline core/src/ldml/ldml_processor.cpp:229]: if the context is empty (see prior loop/comment), pop_back() understandably throws

File: core/src/ldml/ldml_processor.cpp [hunk 2]
```diff
@@ -225,77 +225,84 @@ ldml_processor::process_event(
...
-        // TODO-LDML: unroll ctxt into a str
-        std::u16string ctxtstr;
-        for (size_t i = 0; i < ctxt.size(); i++) {
-          ctxtstr.append(ctxt[i]);
+
+        // Look up the key
+        const std::u16string str = keys.lookup(vk, modifier_state);
+        if (str.empty()) {
+          // not found
+          break; // ----- commit and exit
```
Comment (srl295) [inline core/src/ldml/ldml_processor.cpp:264]: there's probably a better structure for this at this point.  but at least, this is updated to have fewer return; points.

File: core/src/ldml/ldml_processor.cpp [hunk 3]
```diff
@@ -225,77 +225,86 @@ ldml_processor::process_event(
...
+      {
+        // adapted from kmx_processor.cpp
+
+        /** a copy of the current/changed context, for transform use */
+        ldml::string_list ctxt;
+
+        // Construct a context buffer of all the KM_KBP_BT_CHAR items
+        // Extract the context into 'ctxt' for transforms to process
+        if (!!transforms) {
+          // if no transfors, no reason to do this extraction
```
Comment (mcdurdin) [inline core/src/ldml/ldml_processor.cpp:242]:
```suggestion
          // if no transforms, no reason to do this extraction
```

File: core/src/ldml/ldml_processor.cpp [hunk 4]
```diff
@@ -225,77 +225,86 @@ ldml_processor::process_event(
...
+          auto &cp = state->context();
+          // We're only interested in as much of the context as is a KM_KBP_BT_CHAR.
+          uint8_t last_type = KM_KBP_BT_UNKNOWN;
+          for (auto c = cp.rbegin(); c != cp.rend(); c++) {
+            last_type = c->type;
+            if (last_type != KM_KBP_BT_CHAR) {
```
Comment (mcdurdin) [inline core/src/ldml/ldml_processor.cpp:248]: Does this cope with markers interleaved in the context?

**Review by mcdurdin — APPROVED** [review]
I think LGTM, just one question around backspace actions and markers

File: core/src/ldml/ldml_processor.cpp (hunk 2 above)
Comment (mcdurdin) [reply core/src/ldml/ldml_processor.cpp:266]: I don't actually mind `return` -- especially if it reduces depth of algorithm.

File: core/src/ldml/ldml_processor.cpp (hunk 4 above)
Comment (srl295) [reply core/src/ldml/ldml_processor.cpp:248]: It just stops when it gets to a non char (such as a marker). will need to revisit w/ markers

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.129-alpha](https://keyman.com/downloads/releases/alpha/17.0.129)

**Pull request merged**